front/lib/resources/resource_with_vault.ts

```diff
@@ -0,0 +1,128 @@
+import type {
+  Attributes,
+  FindOptions,
+  ForeignKey,
+  Includeable,
+  ModelStatic,
+  NonAttribute,
+  WhereOptions,
+} from "sequelize";
+import { Model } from "sequelize";
+
+import type { Authenticator } from "@app/lib/auth";
+import type { Workspace } from "@app/lib/models/workspace";
+import { BaseResource } from "@app/lib/resources/base_resource";
+import type { VaultModel } from "@app/lib/resources/storage/models/vaults";
+import { VaultResource } from "@app/lib/resources/vault_resource";
+
+// Interface to enforce workspaceId and vaultId.
+interface ModelWithVault {
+  workspaceId: ForeignKey<Workspace["id"]>;
+  vaultId: ForeignKey<VaultModel["id"]>;
+  vault: NonAttribute<VaultModel>;
+}
+
+export type NonAttributeKeys<M> = {
+  [K in keyof M]: M[K] extends NonAttribute<Model<any, any>> ? K : never;
+}[keyof M] &
+  string;
+
+type InferIncludeType<M> = {
+  [K in NonAttributeKeys<M>]: M[K] extends NonAttribute<infer T>
+    ? T extends Model<any, any>
+      ? T
+      : never
+    : never;
+};
+
+export type TypedIncludeable<M> = {
+  [K in NonAttributeKeys<M>]: {
+    model: ModelStatic<InferIncludeType<M>[K]>;
+    as: K;
+  };
+}[NonAttributeKeys<M>];
+
+export interface ResourceFindOptions<M extends Model> {
+  includes?: TypedIncludeable<M>[];
+  limit?: number;
+  order?: FindOptions<M>["order"];
+  where?: WhereOptions<M>;
+}
+
+export abstract class ResourceWithVault<
+  M extends Model & ModelWithVault,
+> extends BaseResource<M> {
+  protected constructor(
+    model: ModelStatic<M>,
+    blob: Attributes<M>,
+    public readonly vault: VaultResource
+  ) {
+    super(model, blob);
+  }
+
+  protected static async baseFetch<
+    T extends ResourceWithVault<M>,
+    M extends Model & ModelWithVault,
+    IncludeType extends Partial<InferIncludeType<M>>,
+  >(
+    this: {
+      new (
+        model: ModelStatic<M>,
+        blob: Attributes<M>,
+        vault: VaultResource,
+        includes?: IncludeType
+      ): T;
+    } & { model: ModelStatic<M> },
+    auth: Authenticator,
+    { includes, limit, order, where }: ResourceFindOptions<M> = {}
+  ): Promise<T[]> {
+    const includeClauses: Includeable[] = [
+      {
+        model: VaultResource.model,
+        as: "vault",
+      },
+      ...(includes || []),
+    ];
+
+    const blobs = await this.model.findAll({
+      where: {
+        ...where,
+        workspaceId: auth.getNonNullableWorkspace().id,
+      } as WhereOptions<M>,
+      include: includeClauses,
+      limit,
+      order,
+    });
+
+    return blobs.map((b) => {
+      const vault = new VaultResource(VaultResource.model, b.vault.get());
```

Comment:
why not including the vault in other includedResults ? you dont need specific handling for the vault here, and that would simplify the constructors. Actually you could even have a more generic baseFetch outside of this file, which do not add vault, but just takes the list of includes

Reply:
The reason vault is not in `includedResults` is because here vault is a resource whereas `includedResults` contains a raw model from Sequelize.

Reply:
ok, makes sense